### ISSUE TITLE
No match in record selector hsTyCon when exporting `JSString -> JSString`

### ISSUE THREAD

**Issue by igrep** [tweag/asterius#102]
# How to reproduce

## Minimum hs file to reproduce

```haskell
import           Asterius.Types        (JSString (JSString), JSVal (JSVal))

foreign export javascript "bug" bug :: JSString -> JSString

bug :: JSString -> JSString
bug = id

main :: IO ()
main = return ()
```

## Build command and result

```
> sudo docker run -it -v `pwd`:/mirror terrorjack/asterius bash -c 'cd /mirror/ ; ahc-link --export-function bug --input-hs ahc-bug.hs'
[INFO] Compiling ahc-bug.hs to WebAssembly
ahc: panic! (the 'impossible' happened)
  (GHC version 8.9.20190403 for x86_64-unknown-linux):
        No match in record selector hsTyCon
```

# NOTE

As long as I grepped the source, [asterius/src/Asterius/JSFFI.hs:530](https://github.com/tweag/asterius/blob/9ea5c51df9076cdb5b76d23bdb17d2715d107a60/asterius/src/Asterius/JSFFI.hs#L530) causes the panic:
`marshalToFFIValueType` seems to handle `JSString` as a `FFI_JSVAL`, which doesn't have `hsTyCon`. So, properly handling `FFI_JSVAL` at JSFFI.hs:530 would fix the problem.
I guess the right `hsTyCon` for `FFI_JSVAL` is `StablePtr`, right?

**Comment by TerrorJack** [comment]
It should already be supported, after all a `JS*` thing is just a `JSVal` under the hood. Thanks for catching this one. Mind opening a PR for it?

**Comment by igrep** [comment]
I haven't tried to build ahc from scratch yet, so I'll try building by myself tomorrow!

**Comment by TerrorJack** [comment]
Won't spoil your fun then :) Let me know when you stumble upon any new trouble. As said in the docs (if I indeed did..), all `JS*` types are a newtype of `JSVal` and it's a special kind of `StablePtr` on the Haskell heap, so the `rts_mk`/`rts_get` functions should use `StablePtr`.

**Comment by igrep** [comment]
I guess in this case, `stack build asterius` on the docker image after fixing will work.
So, I don't have to rebuild ahc (the fork of GHC) right?

**Comment by TerrorJack** [comment]
Yeah, `stack build asterius` should work, otherwise it should be a bug.

`ahc` is an executable in the `asterius` package which serves as a `ghc` wrapper; the `asterius` package is independent from the forked ghc it relies on (much like ghcjs). We have separate build scripts on `circleci-ghc-bindist` branch of this repo to build the forked ghc and distribute as circleci artifacts, then the artifact urls are included in `stack.yaml` of this project, so a plain `stack build` for normal users won't need to rebuild ghc.
